diag_manager/fms_diag_axis_object.F90

```diff
@@ -0,0 +1,287 @@
+
+  end function
+
+  !> @brief Set the subaxis of the axis obj
+  subroutine set_subaxis(obj, bounds)
+    class(diag_axis_t), INTENT(INOUT) :: obj       !< diag_axis obj
+    class(*),           INTENT(INOUT) :: bounds(:) !< bound of the subaxis
+
+    integer :: i !< For do loops
+
+    !< Check if the subaxis for this bouds already exists
+    do i = 1, obj%nsubaxis
+      if (obj%subaxis(i)%exists(bounds)) return
+    enddo
+
+    !< TO DO: everything
+    obj%nsubaxis = obj%nsubaxis + 1
+  end subroutine
+
+  !!!!!!!!!!!!!!!!!! SUB AXIS PROCEDURES !!!!!!!!!!!!!!!!!
+  !> @brief Check if a subaxis was already defined
+  !> @return Flag indicating if a subaxis is already defined
+  function check_if_subaxis_exists(obj,bounds) &
+  result(exists)
+    class(subaxis_t), INTENT(INOUT) :: obj       !< diag_axis obj
+    class(*),         INTENT(IN)    :: bounds(:) !< bounds of the subaxis
+    logical                         :: exists
+
+    !< TO DO: compare bounds
+    exists = .false.
+  end function
+
+  !> @brief Get the length of a 2D domain
+  !> @return Length of the 2D domain
+  function get_length(obj, cart_axis, domain_position) &
+  result (length)
+    class(fms_domain_t), INTENT(INOUT) :: obj       !< diag_axis obj
+    character(len=*),    INTENT(IN)    :: cart_axis !< cart_axis of the axis
+    integer,             INTENT(IN)    :: domain_position !< Domain position (CENTER, NORTH, EAST)
+
+    integer :: length
+
+    select type (obj)
+    type is(fms_domain2d_t)
+      if (trim(cart_axis) == "X") call mpp_get_compute_domain(obj%Domain2, xsize=length, position=domain_position)
+      if (trim(cart_axis) == "Y") call mpp_get_compute_domain(obj%Domain2, ysize=length, position=domain_position)
+    end select
+  end function
+
+  !!!!!!!!!!!!!!!!! FMS_DOMAIN PROCEDURES !!!!!!!!!!!!!!!!!
+
+  !> @brief Set the axis domain
+  subroutine set_axis_domain(obj, Domain, Domain2, DomainU)
+    class(fms_domain_t) :: obj !< fms_domain obj
+    TYPE(domain1d),     INTENT(in),  OPTIONAL :: Domain  !< 1d domain
+    TYPE(domain2d),     INTENT(in),  OPTIONAL :: Domain2 !< 2d domain
+    TYPE(domainUG),     INTENT(in),  OPTIONAL :: DomainU !< Unstructured domain
+
+    select type(obj)
+    type is (fms_domain1d_t)
+      obj%Domain = Domain
+    type is (fms_domain2d_t)
+      obj%Domain2 = Domain2
+    type is (fms_domainUG_t)
+      obj%DomainUG = DomainU
+    end select
+  end subroutine set_axis_domain
+
+  !!!!!!!!!!!!!!!!! OTHER FUNCTIONS/SUBROUTINES !!!!!!!!!!!!!!!!!
+  !> @brief Copy a allocatable string
+  !> @return Allocated string
+  !> TO DO Move this somewhere else?
+  function diag_copy_string(string_in) &
+  result(string_out)
+    character(len=*),             INTENT(IN) :: string_in !< String to copy
+    character(len=:), ALLOCATABLE            :: string_out
+
+    allocate(character(len=len_trim(string_in)) :: string_out)
+    string_out = trim(string_in)
+
+  end function
```

Comment:
1. Why do you need this at all?  Why not just allocate and copy the string in line?
2. There is a whole module of string routines external to diag manager where this would fit better.

Reply:
it is cleaner and less error prone to do:
```F90
string_out = diag_copy_string(string_in)
```

than

```F90
    allocate(character(len=len_trim(string_in)) :: string_out)
    string_out = trim(string_in)
```

especially when it is done multiple times. 

Moving this function is already in the TO DO, it has to be done through main first ...

Reply:
```fortran
string_out = trim(string_in)
```
You shouldn't need to allocate the string length, the assignment should handle that for you.  Then you don't need a function (which should be named `fms_string_copy`) or an allocate statement.  
https://fortranwiki.org/fortran/files/character_handling_in_Fortran.html